store/streaming/plugins/abci/v1/README.md

```diff
@@ -0,0 +1,391 @@
+  repeated cosmos.base.store.v1beta1.StoreKVPair change_set   = 3;
+}
+
+service ABCIListenerService {
+  rpc ListenBeginBlock(ListenBeginBlockRequest) returns (Empty);
+  rpc ListenEndBlock(ListenEndBlockRequest) returns (Empty);
+  rpc ListenDeliverTx(ListenDeliverTxRequest) returns (Empty);
+  rpc ListenCommit(ListenCommitRequest) returns (Empty);
+}
+```
+
+### Generating the Code
+
+To generate the stubs the local client implementation can call, run the following command:
+
+Download required proto files:
+
+```shell
+buf export buf.build/cosmos/cosmos-sdk:${commit} --output store/streaming/plugins/proto
+```
+where `${commit}` is the commit of the buf commit of version of the Cosmos SDK you are using.
+That commit can be found [here](https://github.com/cosmos/cosmos-sdk/blob/main/proto/README.md).
+
+```shell
+protoc -I store/streaming/plugins/proto --go_out=plugins=grpc:. store/streaming/plugins/proto/abci/v1/grpc.proto \
+  && cp -r github.com/cosmos/cosmos-sdk/store/streaming/* store/streaming \
+  && rm -rf github.com 
+```
+
+For other languages you'll need to [download](https://github.com/cosmos/cosmos-sdk/blob/main/third_party/proto/README.md)
+the CosmosSDK protos and the [store/streaming/plugins/proto/abci/v1/grpc.proto](https://github.com/cosmos/cosmos-sdk/blob/main/store/streaming/plugins/proto/abci/v1/grpc.proto)
+into your project and compile. For language specific compilation instructions visit
+[https://github.com/grpc](https://github.com/grpc) and look in the `examples` folder of your
+language of choice `https://github.com/grpc/grpc-{language}/tree/master/examples` and [https://grpc.io](https://grpc.io)
+for the documentation.
+
+### gRPC Client and Server
+
+Implementing the ABCIListener gRPC client and server is a simple and straight forward process.
+
+To create the client and server we create a `ListenerGRPCPlugin` struct that implements the
+`plugin.GRPCPlugin` interface and a `Impl` property that will contain a concrete implementation
+of the `ABCIListener` plugin written in Go.
+
+#### The Interface
+
+The `BaseApp` `ABCIListener` interface will be what will define the plugins capabilities.
+
+Boilerplate RPC implementation of the `ABCIListener` interface. ([store/streaming/plugins/abci/v1/grpc.go](grpc.go))
+```go
+...
+
+var (
+	_ storetypes.ABCIListener = (*GRPCClient)(nil)
+)
+
+// GRPCClient is an implementation of the ABCIListener interface that talks over gRPC.
+type GRPCClient struct {
+	client ABCIListenerServiceClient
+}
+
+func (m GRPCClient) ListenBeginBlock(ctx context.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error {
+	_, err := m.client.ListenBeginBlock(ctx, &ListenBeginBlockRequest{
+		Req: &req,
+		Res: &res,
+	})
+	return err
+}
+
+func (m GRPCClient) ListenEndBlock(ctx context.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error {
+	_, err := m.client.ListenEndBlock(ctx, &ListenEndBlockRequest{
+		Req: &req,
+		Res: &res,
+	})
+	return err
+}
```

Comment:
But wait, these Listen methods mutate stuff in the Res field they're given, right? But the res provided to this client's ListenWhateverBlock is a value type, which means even if you call client.ListenWhatever with a `&res` address of that input param, whatever changes the ListenWhateverBlock method(s) make to that parameter will not be visible to the caller of this GRPCClient method. Those changes are applied to the _copy of_ the `res` value passed to the method only.

This is a pretty straightforward bug that should have been caught by unit tests -- where is this method tested?

Reply:
Why would we want to make changes to the req, res?

Reply:
> But wait, these Listen methods mutate stuff in the Res field they're given, right? But the res provided to this client's ListenWhateverBlock is a value type, which means even if you call client.ListenWhatever with a `&res` address of that input param, whatever changes the ListenWhateverBlock method(s) make to that parameter will not be visible to the caller of this GRPCClient method. Those changes are applied to the _copy of_ the `res` value passed to the method only.
> 
> This is a pretty straightforward bug that should have been caught by unit tests -- where is this method tested?

These are listeners; they shouldn't be changing anything. And as you pointed out, they can't.

Reply:
If they're not changing anything, then it's confusing that they take req and res by reference, given the outer method here receives them by value. But there's probably an obscure reason for it, I guess.

Reply:
All Go generated gRPC code uses references for the client/server methods. I'm guessing the reason has to do with protoreflection. [#3 on this blog post](https://go.dev/blog/laws-of-reflection) may shed some light to the reason.